packages/bloc_test/lib/src/bloc_test.dart

```diff
@@ -229,26 +229,36 @@ Future<void> testBloc<B extends BlocBase<State>, State>({
       await subscription.cancel();
       await verify?.call(bloc);
       await tearDown?.call();
-    },
-    (Object error, _) {
-      if (shallowEquality && error is test.TestFailure) {
-        // ignore: only_throw_errors
-        throw test.TestFailure(
-          '''
+    });
+  } catch (error) {
+    if (shallowEquality && error is test.TestFailure) {
+      // ignore: only_throw_errors
+      throw test.TestFailure(
+        '''
 ${error.message}
 WARNING: Please ensure state instances extend Equatable, override == and hashCode, or implement Comparable.
 Alternatively, consider using Matchers in the expect of the blocTest rather than concrete state instances.\n''',
-        );
-      }
-      if (errors == null || !unhandledErrors.contains(error)) {
-        // ignore: only_throw_errors
-        throw error;
-      }
-    },
-  );
+      );
+    }
+    if (errors == null || !unhandledErrors.contains(error)) {
+      rethrow;
+    }
+  }
+
   if (errors != null) test.expect(unhandledErrors, test.wrapMatcher(errors()));
 }
 
+Future<void> _runZonedGuarded(Future<void> Function() body) {
+  final completer = Completer<void>();
+  runZonedGuarded(() async {
+    await body();
+    if (!completer.isCompleted) completer.complete();
+  }, (error, stack) {
+    if (!completer.isCompleted) completer.completeError(error);
```

Comment:
`completeError()` can take the `stack` too as the 2nd param. I think it would be worthwile to pass if its already there.

Reply:
Yeah good call this pr isn’t done yet, still need to add tests and I’ll update to include stackTrace as well 👍